packages/jaeger-ui/src/model/link-patterns.js

```diff
@@ -0,0 +1,183 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import _uniq from 'lodash/uniq';
+import { getConfigValue } from '../utils/config/get-config';
+
+const parameterRegExp = /\$\{([^{}]*)\}/;
+
+export function processTemplate(template, encodeFn) {
+  if (typeof template !== 'string') {
+    if (!template || !Array.isArray(template.parameters) || !(template.template instanceof Function)) {
+      throw new Error('Invalid template');
+    }
+    return template;
+  }
+  const templateSplit = template.split(parameterRegExp);
+  const templateSplitLength = templateSplit.length;
+  const parameters = [];
+  // odd indexes contain variable names
+  for (let i = 1; i < templateSplitLength; i += 2) {
+    const param = templateSplit[i];
+    let paramIndex = parameters.indexOf(param);
+    if (paramIndex === -1) {
+      paramIndex = parameters.length;
+      parameters.push(param);
+    }
+    templateSplit[i] = paramIndex;
+  }
+  return {
+    parameters,
+    template: (...args) => {
+      let text = '';
+      for (let i = 0; i < templateSplitLength; i++) {
+        if (i % 2 === 0) {
+          text += templateSplit[i];
+        } else {
+          text += encodeFn(args[templateSplit[i]]);
+        }
+      }
+      return text;
+    },
+  };
+}
+
+export function createTestFunction(entry) {
+  if (typeof entry === 'string') {
+    return arg => arg === entry;
+  }
+  if (Array.isArray(entry)) {
+    return arg => entry.indexOf(arg) > -1;
+  }
+  if (entry instanceof RegExp) {
+    return arg => entry.test(arg);
+  }
+  if (typeof entry === 'function') {
+    return entry;
+  }
+  if (entry == null) {
+    return () => true;
+  }
+  throw new Error(`Invalid value: ${entry}`);
+}
+
+const identity = a => a;
+
+export function processLinkPattern(pattern) {
+  try {
+    const url = processTemplate(pattern.url, encodeURIComponent);
+    const text = processTemplate(pattern.text, identity);
+    return {
+      object: pattern,
+      type: createTestFunction(pattern.type),
+      key: createTestFunction(pattern.key),
+      value: createTestFunction(pattern.value),
+      url,
+      text,
+      parameters: _uniq(url.parameters.concat(text.parameters)),
+    };
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error(`Ignoring invalid link pattern: ${error}`, pattern);
+    return null;
+  }
+}
+
+export function getParameterInArray(name, array) {
+  if (array) {
+    return array.find(entry => entry.key === name);
+  }
+  return undefined;
+}
+
+export function getParameterInAncestor(name, spans, startSpanIndex) {
+  let currentSpan = { depth: spans[startSpanIndex].depth + 1 };
+  for (let spanIndex = startSpanIndex; spanIndex >= 0; spanIndex--) {
```

Comment:
This will not exit early when a root level ancestor is encountered, e.g.

```js
for (...) {
  ...
  if (currentSpan.depth === 0) {
    break; // as there will be no more ancestors
  }
}
```

Reply:
I have now added this condition in the for loop to exit early when reaching a root level ancestor.